src/components/tl/ucp/alltoall/alltoall_onesided.c

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "alltoall.h"
+#include "core/ucc_progress_queue.h"
+#include "utils/ucc_math.h"
+#include "tl_ucp_sendrecv.h"
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_progress(ucc_coll_task_t *ctask);
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_start(ucc_coll_task_t *ctask)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team = TASK_TEAM(task);
+    ucc_status_t       status;
+
+    task->super.super.status = UCC_INPROGRESS;
+    status = ucc_tl_ucp_alltoall_onesided_progress(&task->super);
+    if (UCC_INPROGRESS == status) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+    task->super.super.status = status;
+    ucc_task_complete(ctask);
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_progress(ucc_coll_task_t *ctask)
+{
+    ucc_tl_ucp_task_t *task   = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team   = TASK_TEAM(task);
+    ptrdiff_t          src    = (ptrdiff_t)TASK_ARGS(task).src.info.buffer;
+    ptrdiff_t          dest   = (ptrdiff_t)TASK_ARGS(task).dst.info.buffer;
+    size_t             nelems = TASK_ARGS(task).src.info.count;
+    ucc_rank_t         grank  = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         gsize  = UCC_TL_TEAM_SIZE(team);
+    ucc_rank_t         start  = (grank + 1) % gsize;
+    long *             pSync;
+    ucc_rank_t         peer;
+
+    /* TODO: change when support for library-based work buffers is complete */
+    pSync = TASK_ARGS(task).global_work_buffer;
+
+    if (task->send_posted > 0) {
+        if (*pSync < gsize - 1 || task->send_completed < task->send_posted) {
+            ucp_worker_progress(UCC_TL_UCP_TEAM_CTX(team)->ucp_worker);
+            return UCC_INPROGRESS;
+        }
+        goto completed;
+    }
+
+    nelems = (nelems / gsize) * ucc_dt_size(TASK_ARGS(task).src.info.datatype);
+    dest   = dest + grank * nelems;
+    ucc_tl_ucp_put_nb((void *)(src + start * nelems), (void *)dest, nelems,
+                      start, team, task);
+    ucc_tl_ucp_atomic_inc(pSync, start, team, task);
+
+    for (peer = (start + 1) % gsize; peer != start; peer = (peer + 1) % gsize) {
```

Comment:
i think it's better to move this loop and ucc_tl_ucp_put_nb above to ucc_tl_ucp_alltoall_onesided_start because anyway that code will be executed only once at collective start. So we can simplify progress function and reduce number of additional checks in it